File: numberOfBoomerangs/numberOfBoomerangs.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int numberOfBoomerangs(vector<pair<int, int>>& points) {
        int res = 0;
        if (points.size() < 3)
            return res;
        unordered_map<int64_t, int> m;
        for (int i = 0; i < points.size(); i++) {
            for (int j = 0; j < points.size(); j++) {
                if (j == i)
                    continue;
                int dist_x = points[j].first - points[i].first;
                int dist_y = points[j].second - points[i].second;
                int64_t dist = dist_x * dist_x + dist_y * dist_y;
                m[dist]++;
            }
            for (auto& p : m) {
                res += p.second * (p.second - 1);
            }
            m.clear();
        }
        return res;
    }
};

int main()
{
    Solution s;
    vector<pair<int, int> > vec;
    vec.push_back(make_pair(0,0));
    vec.push_back(make_pair(1,0));
    vec.push_back(make_pair(2,0));
    cout << s.numberOfBoomerangs(vec);
    cout << endl;
    return 0;
}

File: numberOfBoomerangs/numberOfBoomerangs.java
class Solution {
    public int numberOfBoomerangs(int[][] points) {
        int res = 0;
        int pNumber = points.length;
        Map<Integer, Integer> map = new HashMap<>(); // distince -> number
        for (int i = 0; i < pNumber; i++) {
            for (int j = 0; j < pNumber; j++) {
                if (i == j) continue;
                int d = getDistance(points[i], points[j]);
                map.put(d, map.getOrDefault(d, 0) + 1);
            }
            for (int val : map.values()) {
                res += val * (val - 1);
            }
            map.clear();
        }
        return res;
    }
    
    public int getDistance(int[] p1, int[] p2) {
        int dx = p1[0] - p2[0];
        int dy = p1[1] - p2[1];
        return dx * dx + dy * dy;
    }
}
